### PR TITLE
[CSL-511] Expose error code in network request failures

### PR DESCRIPTION
**Pull request by esezen**

### CODE REVIEW

**Review by sblaurock — COMMENTED** [review]
Tests please 😄

File: constructorio-client/src/main/java/io/constructor/client/ConstructorException.java [hunk 1]
```diff
@@ -1,11 +1,28 @@
...
+
 	public ConstructorException(String msg) {
 		super(msg);
 	}
 
 	public ConstructorException(Exception e) {
 		super(e);
+		if(e instanceof ConstructorException) {
+			Integer errorCode = ((ConstructorException) e).getErrorCode();
+			if(errorCode != null){
```
Comment (sblaurock) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorException.java:14]: `if (`

**Review by sblaurock — COMMENTED** [review]
This looks good to me! @rubikzube could you have a final look?

**Review by rubikzube — CHANGES_REQUESTED** [review]

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 2]
```diff
@@ -1088,20 +1088,26 @@ protected Builder makeUserRequestBuilder(UserInfo info) {
...
             errorMessage = "[HTTP " + response.code() + "]";
         } finally {
             response.close();
         }
-        throw new ConstructorException(errorMessage);
+        if(errorCode != null) {
+            throw new ConstructorException(errorMessage, errorCode);
+        } else {
+            throw new ConstructorException(errorMessage);
+        }
```
[the remark this answers is not in the document]
Comment (esezen) [reply constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java]: I thought this would cause an issue but it did not. Thanks! Updated

**Review by sblaurock — APPROVED** [review]

File: constructorio-client/src/main/java/io/constructor/client/ConstructorException.java [hunk 3]
```diff
@@ -1,11 +1,28 @@
...
 		super(msg);
 	}
 
 	public ConstructorException(Exception e) {
 		super(e);
+		if (e instanceof ConstructorException) {
+			Integer errorCode = ((ConstructorException) e).getErrorCode();
+			if (errorCode != null){
+				this.errorCode = errorCode;
+			}
```
Comment (rubikzube) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorException.java:16]: This could also be simplified ...

Comment (rubikzube) [reply constructorio-client/src/main/java/io/constructor/client/ConstructorException.java:16]:
```java
Integer errorCode = ((ConstructorException) e).getErrorCode();
if (errorCode != null){
   this.errorCode = errorCode;
}
```
is the same as

```java
this.errorCode = ((ConstructorException) e).getErrorCode();
```

Comment (esezen) [reply constructorio-client/src/main/java/io/constructor/client/ConstructorException.java:16]: As you can tell I am scared off nulls in Java

**Review by rubikzube — APPROVED** [review]

**Pull request merged**